packages/backend/src/core/preprocessing/HistoryPreprocessor.ts

```diff
@@ -0,0 +1,108 @@
+import { AssetHash, AssetId } from '@explorer/types'
+import { Knex } from 'knex'
+
+import {
+  PreprocessedAssetHistoryRecord,
+  PreprocessedAssetHistoryRepository,
+} from '../../peripherals/database/PreprocessedAssetHistoryRepository'
+import { StateUpdateRecord } from '../../peripherals/database/StateUpdateRepository'
+import { Logger } from '../../tools/Logger'
+
+export abstract class HistoryPreprocessor<T extends AssetHash | AssetId> {
+  constructor(
+    protected preprocessedAssetHistoryRepository: PreprocessedAssetHistoryRepository<T>,
+    protected logger: Logger
+  ) {
+    this.logger = this.logger.for(this)
+  }
+
+  abstract preprocessNextStateUpdate(
+    trx: Knex.Transaction,
+    stateUpdate: StateUpdateRecord
+  ): Promise<void>
+
+  // Closing a position or vault is a special case because the state update
+  // has StarkKey set to ZERO and all assets empty, so we need to retrieve it
+  // all from history
+  async closePositionOrVault(
+    trx: Knex.Transaction,
+    positionId: bigint,
+    stateUpdate: StateUpdateRecord,
+    assetPrices: Record<string, bigint>
+  ) {
+    const recordsToClose =
+      await this.preprocessedAssetHistoryRepository.getCurrentNonEmptyByPositionOrVaultId(
```

Comment:
Why only not empty?

Reply:
There was a reason but I found a way to make things even simpler and we no longer need to fetch non-empty here. Thanks :)